File: exercises/ddd_infrastructure_layer/typescript/src/domain/application/Application.ts
import type { ApplicationId } from "./ApplicationId";
import type { ApplicationReceivedDateTime } from "./ApplicationReceivedDateTime";
import { ApplicationStage } from "./ApplicationStage";
import { ApplicationStatus } from "./ApplicationStatus";
import type { EmailAddress } from "./EmailAddress";
import type { FullName } from "./FullName";
import type { MotivationLetter } from "./MotivationLetter";
import type { PhoneNumber } from "./PhoneNumber";

export class Application {
  private _id: ApplicationId;
  private _stage: ApplicationStage = ApplicationStage.Zero;
  private _status: ApplicationStatus = ApplicationStatus.InProgress;
  private _receivedDateTime: ApplicationReceivedDateTime;
  private _fullName?: FullName;
  private _emailAddress?: EmailAddress;
  private _phoneNumber?: PhoneNumber;
  private _motivationLetter?: MotivationLetter;

  get id() { return this._id; }
  get receivedDateTime() { return this._receivedDateTime; }
  get fullName() { return this._fullName; }
  get emailAddress() { return this._emailAddress; }
  get phoneNumber() { return this._phoneNumber; }
  get motivationLetter() { return this._motivationLetter; }
  get stage() { return this._stage; }
  get status() { return this._status; }

  constructor(id: ApplicationId, receivedDateTime: ApplicationReceivedDateTime) {
    if (!id || !receivedDateTime) {
      throw new Error("id and receivedDateTime are required");
    }
    this._id = id;
    this._receivedDateTime = receivedDateTime;
  }

  fillOutBasicDetails(fullName: FullName, emailAddress: EmailAddress, phoneNumber: PhoneNumber) {
    const newFields = { ...this.fields(), fullName, emailAddress, phoneNumber };
    if (!ApplicationStage.One.isValidForStage(newFields)) {
      throw new Error("Invalid data or application not in correct stage");
    }
    this._fullName = fullName;
    this._emailAddress = emailAddress;
    this._phoneNumber = phoneNumber;
    this._stage = ApplicationStage.One;
  }

  submitMotivationLetter(motivationLetter: MotivationLetter) {
    const newFields = { ...this.fields(), motivationLetter };
    if (!ApplicationStage.Two.isValidForStage(newFields)) {
      throw new Error("Invalid data or application not in correct stage");
    }
    this._motivationLetter = motivationLetter;
    this._stage = ApplicationStage.Two;
  }

  approveInitialApplication() {
    if (!ApplicationStage.Three.isValidForStage(this.fields())) {
      throw new Error("Application not in correct stage to approve");
    }
    this._stage = ApplicationStage.Three;
  }

  reject() {
    this._status = ApplicationStatus.Rejected;
  }

  equals(other: Application) {
    return this._id.equals(other._id);
  }

  private fields(): Record<string, any> {
    return {
      id: this.id,
      receivedDateTime: this.receivedDateTime,
      fullName: this.fullName,
      emailAddress: this.emailAddress,
      phoneNumber: this.phoneNumber,
      motivationLetter: this.motivationLetter,
    };
  }
}

File: exercises/ddd_infrastructure_layer/java/src/main/java/domain/application/Application.java
package domain.application;
import domain.application.stages.ApplicationStage;
import domain.application.stages.ApplicationStages;

public class Application {
    private ApplicationId id;
    private ApplicationStage stage = ApplicationStages.ZERO;
    private ApplicationStatus status = ApplicationStatus.IN_PROGRESS;
    private ApplicationReceivedDateTime receivedDateTime;
    private FullName fullName;
    private EmailAddress emailAddress;
    private PhoneNumber phoneNumber;
    private MotivationLetter motivationLetter;

    public Application(ApplicationId id, ApplicationReceivedDateTime receivedDateTime) {
        var fields = new ApplicationFields(id, stage, status, receivedDateTime, null, null, null, null);
        if (!ApplicationStages.ZERO.isValidForStage(fields)) {
            throw new IllegalArgumentException("Invalid data for initial application");
        }
        this.id = id;
        this.receivedDateTime = receivedDateTime;
    }

    public ApplicationId getId() { return id; }
    public ApplicationStage getStage() { return stage; }
    public ApplicationStatus getStatus() { return status; }
    public ApplicationReceivedDateTime getReceivedDateTime() { return receivedDateTime; }
    public FullName getFullName() { return fullName; }
    public EmailAddress getEmailAddress() { return emailAddress; }
    public PhoneNumber getPhoneNumber() { return phoneNumber; }
    public MotivationLetter getMotivationLetter() { return motivationLetter; }

    public void fillOutBasicDetails(FullName fullName, EmailAddress emailAddress, PhoneNumber phoneNumber) {
        var fields = new ApplicationFields(
            this.id,
            this.stage,
            this.status,
            this.receivedDateTime,
            fullName,
            emailAddress,
            phoneNumber,
            this.motivationLetter
        );
        if (!ApplicationStages.ONE.isValidForStage(fields)) {
            throw new IllegalArgumentException("Invalid data or application not in correct stage");
        }
        this.fullName = fullName;
        this.emailAddress = emailAddress;
        this.phoneNumber = phoneNumber;
        this.stage = ApplicationStages.ONE;
    }

    public void fillOutMotivationLetter(MotivationLetter letter) {
        var fields = new ApplicationFields(
            this.id,
            this.stage,
            this.status,
            this.receivedDateTime,
            this.fullName,
            this.emailAddress,
            this.phoneNumber,
            letter
        );
        if (!ApplicationStages.TWO.isValidForStage(fields)) {
            throw new IllegalArgumentException("Invalid data or application not in correct stage");
        }
        this.motivationLetter = letter;
        this.stage = ApplicationStages.TWO;
    }

    public void approveInitialApplication() {
        var fields = new ApplicationFields(
            this.id,
            this.stage,
            this.status,
            this.receivedDateTime,
            this.fullName,
            this.emailAddress,
            this.phoneNumber,
            this.motivationLetter
        );
        if (!ApplicationStages.THREE.isValidForStage(fields)) {
            throw new IllegalArgumentException("Invalid data or application not in correct stage");
        }
        this.stage = ApplicationStages.THREE;
    }

    public void reject() {
        this.status = ApplicationStatus.REJECTED;
    }


    public boolean equals(Application other) {
        return this.id.equals(other.id);
    }

    public record ApplicationFields(
        ApplicationId id,
        ApplicationStage stage,
        ApplicationStatus status,
        ApplicationReceivedDateTime receivedDateTime,
        FullName fullName,
        EmailAddress emailAddress,
        PhoneNumber phoneNumber,
        MotivationLetter motivationLetter
    ) { }
}

